### ISSUE TITLE
Error: EMFILE on OSX when requiring lots of files

### ISSUE THREAD

**Issue by karl** [substack/node-browserify#431]
When requiring 256 files from a single file on OSX browserify throws:

```
Error: EMFILE, open '/Users/karl/Sites/KVP/GuruFramework/package.json'
```

From Googling around it looks like this is caused by attempting to open too many files in parallel.

It looks like this could be solved by using the `graceful-fs` module to limit the number of files opened in parallel.

https://github.com/isaacs/node-graceful-fs

**Comment by freakynit** [comment]
ulimit -n 2560 solved the issue

**Comment by garethbjohnson** [comment]
@karl Thanks for the fix!

**Comment by ADOMAKOR412** [comment]
@karl How did you use graceful-fs as a work around? I'm new to this. I tried copying and pasting your work around to my ~/.bash_profile and it did not work.

**Comment by karl** [comment]
I never actually implemented a `graceful-fs` fix I'm afraid, I just suggested it as a possibility.

I'm still using the ulimit workaround.

**Comment by tmpvar** [comment]
just ran into this as well

**Comment by tschaub** [comment]
I just tried to upgrade to browserify@11.0.1 and started getting `EMFILE` errors on our CI server (Ubuntu 12.04 - which reports `unlimited` for `ulimit`).  Can this issue be retitled to remove the OSX specific part?  Using `graceful-fs` would solve the issue everywhere.

**Comment by programmarchy** [comment]
Make sure you are running `ulimit -n` not just `ulimit` to determine the maximum number of open file descriptors.

**Comment by tschaub** [comment]
> Make sure you are running `ulimit -n` not just `ulimit` to determine the maximum number of open file descriptors.

Thanks @programmarchy, my bad.  So 1024.  I think it is still valid to say that this should not be considered an OS specific issue.

**Comment by RangerMauve** [comment]
Getting this in Windows now and sadly it's a garbage OS and doesn't seem to have a simple way to change the limit.

**Comment by jackcviers** [comment]
@RangerMauve
http://blogs.technet.com/b/markrussinovich/archive/2009/09/29/3283844.aspx

My understanding is that Windows has a limit of somewhere around 16.5
million handles, so upping it might not solve your problem.

Maybe you could `npm dedupe` and try again?

-- 
Jack Viers
333 E. Grand Ave.
#115
Des Moines, IA 50309
515.313.7838

**Comment by RangerMauve** [comment]
@jackcviers I'm not so sure that's true anymore. I only started having this issue after a Windows 10 update earlier in the week. Before then it was all working properly.

Also, it's not so much that I have a lot of dupes, I'm just running several builds at once in a gulp script.

**Comment by jackcviers** [comment]
Here's a possible answer:
http://serverfault.com/questions/108827/windows-file-handle-limit

I know that even one build running through briwserify of a fairly large app
on mac would trigger the handle limit right away.

I know the point of gulp is lost if you just use a single synchronous
build, but for workaround purposes you could start with one and add a build
at a time in parallel until you hit the limit. Then try to create a
sequence of tasks for all your builds, limiting your parallelism to each
one to the amount that runs; e.g build one thru 3 don't depend on anything,
3 thru six depend on one, two, and three, etc. That will be a minor pain in
gulp, but might still give you acceptable throughput without crashing.

Other possible workarounds, asuming w10 handler limit cannot be changed,
are watchify, running your build in a docker / vagrant / environment where
you can set the limit appropriately, or vreaking your project into smaller
pieces with individual builds and common dependencies.

On Fri, Sep 11, 2015, 09:39 RangerMauve <notifications@github.com> wrote:

> @jackcviers <https://github.com/jackcviers> I'm not so sure that's true
> anymore. I only started having this issue after a Windows 10 update earlier
> in the week. Before then it was all working properly.
>
> Also, it's not so much that I have a lot of dupes, I'm just running
> several builds at once in a gulp script.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/substack/node-browserify/issues/431#issuecomment-139563348>
> .
>
-- 
Jack Viers
333 E. Grand Ave.
#115
Des Moines, IA 50309
515.313.7838

**Comment by RangerMauve** [comment]
What I ended up doing was ensuring that the steps run serially rather than in parallel. It makes the build take more time, but at least it works now.

**Comment by freynolds** [comment]
Updating to the latests node.js solved this issue for me!

**Comment by adsteel** [comment]
Thanks @freynolds! Upgrading to node version 5.2 from 0.10 solved my problem, too.

**Comment by jiezhang917** [comment]
The similar issue happen to me when i run gulp with watch, it says events.js throw err without handling and watch null emfile. So i searched emfile error and saw someone saying reboot. I just rebooted my pc and everything is fine now.

**Comment by ViktorHofer** [comment]
Hello,

I'm facing the same issues as @RangerMauve but on Windows 7. This started to happen when my project got bigger with more modules. Changing the ulimit in Windows (if that's possible) is no option because these machines are administrated by the customer.

Is it still planned to use graceful-fs instead of fs? Currently I'm not able to build my project entirely. I have to split it into two parts.

Thanks!

**Comment by RangerMauve** [comment]
Try using npm 3. It's more agressive at keeping the node_modules folder flat.

**Comment by ViktorHofer** [comment]
I'm already using npm 3... And with modules I ment application entry points => bundles for browserify. The problem is that I'M creating multiple bundles at the same time. I need to implement some kind of scheduler which watches over it and limits it e.g. to 5 parallel browserify bundlings...
